ndcube/ndcollection.py

```diff
@@ -0,0 +1,262 @@
+            Default is ("0", "1",...)
+
+        aligned_axes: `tuple` of `int`, `tuple` of `tuple`s of `int`, or None (Optional)
+            Axes of each cube/sequence that are aligned in numpy order.
+            If elements are int, then the same axis numbers in all cubes/sequences are aligned.
+            If elements are tuples of ints, then must be one tuple for every cube/sequence.
+            Each element of each tuple gives the axes of each cube/sequence that are aligned.
+            Default="All", i.e. all axes are aligned.
+
+        meta: `dict` (Optional)
+            General metadata for the overall collection.
+
+        Example
+        -------
+        Say the collection holds two NDCubes, each of 3 dimensions.
+        aligned_axes = (1, 2)
+        means that axis 1 (0-based counting) of cube0 is aligned with axis 1 of cube1,
+        and axis 2 of cube0 is aligned with axis 2 of cube1.
+        However, if
+        aligned_axes = ((0, 1), (2, 1))
+        then the first tuple corresponds to cube0 and the second with cube1.
+        This is interpretted as axis 0 of cube0 is aligned with axis 2 of cube1 while
+        axis 1 of cube0 is aligned with axis 1 of cube1.
+
+        """
+        # Check inputs
+        # Ensure there are no duplicate keys
+        if keys is None:
+            keys = np.arange(len(data)).astype("str")
+        elif len(set(keys)) != len(keys):
+            raise ValueError("Duplicate keys detected.")
+        if len(keys) != len(data):
+            raise ValueError("Data and keys inputs of different lengths.")
+        data_types = [isinstance(cube, SUPPORTED_COLLECTION_TYPES) for cube in data]
+        if not all(data_types):
+            bad_data_index = np.arange(len(data))[np.invert(data_types)][0]
+            raise TypeError(
+                    "All data entries must be of types {0}.\n{1}th data entry is type {2}".format(
+                        SUPPORTED_COLLECTION_TYPES, bad_data_index, type(data[bad_data_index])))
+
+        self._first_key = keys[0]
+        self._cube_types = type(data[0])
+
+        # Enter data into object.
+        super().__init__([(key, cube) for key, cube in zip(keys, data)])
+        self.meta = meta
+
+        n_cubes = len(data)
+        # If aligned_axes not set, assume all axes are aligned in order.
+        if aligned_axes == "All":
+            # Check all cubes are of same shape
+            cube0_dims = data[0].dimensions
+            cubes_same_shape = all(
+                [all(data[i].dimensions == cube0_dims) for i in range(n_cubes)])
+            if cubes_same_shape is not True:
+                raise ValueError(
+                    "All cubes in data not of same shape. Please set aligned_axes kwarg.")
+            self.n_aligned_axes = len(cube0_dims)
+            self.aligned_axes = dict([(keys[i], tuple(range(len(cube0_dims))))
+                                      for i in range(n_cubes)])
```

Comment:
Would this not be better as a `collections.defaultdict` object?

Reply:
No!